[FEATURE] Add KopEventManager to repair metadata consistency of cluster and consistency of consumer group status

wenbingshen opened the pull request:
Fixes #617 

### Motivation
When investigating #617, I found the root cause of TimeOut exception and incorrect consumer group status.

**The first question is a new problem I found while investigating the cause**
**1. For example, when the cluster contains two kop nodes, broker1 is the owner broker of all topics in the current cluster, and broker2 is not the leader of any partition.**

- 1.1. When the client describesGroup, it needs to connect to the configured brokerList first.
- 1.2. Suppose it connects to broker2 to obtain metadata. For the current server-side metadata processing logic, if the leader nodes of all partitions are not broker2, they are all broker1. Since metadata acquisition proceeds normally, the metadata returned by the brokers node only contains broker1.
- 1.3. So next, the client will send FindCoordinatorRequest to broker1. Broker1 calculates based on the consumer group hash and finds that the coordinator is on broker2.
- 1.4. Because the cluster attribute in the client metadata only has the broker1 node, the client can no longer send DescribeGroupRequest to the broker2, and keep sending the update metadata request to broker1 until it times out.

**The second question is described in #617** 
**2. The current GroupCoordinator will not remove the deleted partition from the GroupMetadata after the topic is deleted**

### Modifications
Add a single-threaded event processing queue on the kop server. It will obtain the current cluster node metadata once at startup and place it locally, and will also register to monitor the `/loadbalance/broker` subnode and `/kop/delete_topics` subnode, `/loadbalance /broker` is used to update the metadata of the kop node list, and `/kop/delete_topics` is used to update the GroupMetadata which had these deleted topics, and try to change the state of the consumer group.

wenbingshen commented:
1. When a new node started, the kop brokers node metadata update is triggered.
![image](https://user-images.githubusercontent.com/35599757/132251241-65d40618-de45-4a0b-a93a-2fe5f8007567.png)

wenbingshen commented:
2. Remove the deleted partition from the GroupMetadata and try to update the consumer group status.

![image](https://user-images.githubusercontent.com/35599757/132251840-03c804f7-7eb1-4a1d-8f15-b9d84731a998.png)

wenbingshen commented:
@BewareMyPower There are still some bugs in the code that need to be fixed and related tests added. Before that, can you first take a look at the solution described by this pr? thanks.

wenbingshen commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:
```diff
@@ -0,0 +1,296 @@
+        }
+    }
+
+    private void getBrokers(List<String> pulsarBrokers) {
+        HashSet<Node> kopBrokers = Sets.newHashSet();
+        pulsarBrokers.forEach(broker -> {
+            try {
+                byte[] brokerInfo =
+                        kopZkClient.getDataForPath(KopZkClient.getBrokersChangeZNodePath() + "/" + broker);
+                JsonObject jsonObject = parseJsonObject(new String(brokerInfo));
+                JsonObject protocols = jsonObject.getAsJsonObject("protocols");
+                JsonElement element = protocols.get("kafka");
+
+                if (element != null) {
+                    String kopBrokerStr = element.getAsString();
+                    Node kopNode = getNode(kopBrokerStr);
+                    kopBrokers.add(kopNode);
+                }
+            } catch (Exception e) {
+                log.error("Get broker {} ZNode data failed which have an error {}", broker, e.getMessage());
+                e.printStackTrace();
+            }
+        });
+        Collection<? extends Node> oldKopBrokers = adminManager.getBrokers();
+        adminManager.addBrokers(kopBrokers);
+        log.info("Refresh kop brokers new cache {}, old brokers cache {}",
+                adminManager.getBrokers(), oldKopBrokers);
+    }
```
Get the address of the kop node under the protocol attribute.

wenbingshen commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:
```diff
@@ -0,0 +1,296 @@
+                topicsDeletions.forEach(topic -> {
+                    KopTopic kopTopic = new KopTopic(topic);
+                    kopTopicsSet.add(kopTopic);
+                    topicsFullNameDeletionsSets.add(kopTopic.getFullName());
+                });
+
+                log.debug("Delete topics listener fired for topics {} to be deleted", topicsDeletions);
+                Iterable<GroupMetadata> groupMetadataIterable = coordinator.getGroupManager().currentGroups();
+                HashSet<TopicPartition> topicPartitionsToBeDeletions = Sets.newHashSet();
+
+                groupMetadataIterable.forEach(groupMetadata -> {
+                    topicPartitionsToBeDeletions.addAll(
+                            groupMetadata.collectPartitionsWithTopics(topicsFullNameDeletionsSets));
+                });
+
+                Set<String> deletedTopics = Sets.newHashSet();
+                if (!topicPartitionsToBeDeletions.isEmpty()) {
+                    coordinator.handleDeletedPartitions(topicPartitionsToBeDeletions);
+                    Set<String> collectDeleteTopics = topicPartitionsToBeDeletions
+                            .stream()
+                            .map(TopicPartition::topic)
+                            .collect(Collectors.toSet());
+
+                    deletedTopics = kopTopicsSet.stream().filter(
+                            kopTopic -> collectDeleteTopics.contains(kopTopic.getFullName())
+                    ).map(KopTopic::getOriginalName).collect(Collectors.toSet());
+
+                    kopZkClient.deleteNodesForPaths(
+                            KopZkClient.getDeleteTopicsZNodePath(), deletedTopics);
+                }
+
+                log.info("GroupMetadata delete topics {}, no matching topics {}",
+                        deletedTopics, Sets.difference(topicsFullNameDeletionsSets, deletedTopics));
+
+            } catch (Exception e) {
+                log.error("DeleteTopicsEvent process have an error {}", e.getMessage());
+                e.printStackTrace();
+            }
+        }
+    }
```
Remove partitions in consumer groups for partitions which have been deleted.

wenbingshen commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:
```diff
@@ -0,0 +1,296 @@
+        }
+    }
+
+    class BrokersChangeEvent implements KopEvent {
+        @Override
+        public void process() {
+            try {
+                getBrokers(kopZkClient.getBrokers());
+            } catch (Exception e) {
+                log.error("BrokersChangeEvent process have an error {}", e.getMessage());
+                e.printStackTrace();
+            }
+        }
+    }
```
/loadbalance/broker broker change event.

wenbingshen reviewed (commented):
Processing cluster brokers metadata updates and processing deleted topics for consumer groups

wenbingshen commented:
@BewareMyPower My PR and I are ready. Zk client has been replaced with MetadataStore.  PTAL :)

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:
```diff
@@ -0,0 +1,291 @@
+
+
+    public void put(KopEvent event) {
+        try {
+            putLock.lock();
+            queue.put(event);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            log.error("Error put event {} to coordinator event queue {}", event, e);
+        } finally {
+            putLock.unlock();
+        }
+    }
```
We prefer the following lock idiom.

```java
lock.lock();
try {
    // ...
} finally {
    lock.unlock();
}
```

Because if `lock.lock()` failed with exceptions, `.unlock()` would still be called.
See examples in https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html.

BewareMyPower replied:
The same applies to `clearAndPut`.

wenbingshen replied:
good comment!

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:
```diff
@@ -0,0 +1,291 @@
+                    }
+                }
+            } catch (Exception e) {
+                log.error("Get broker {} path data failed which have an error {}", broker, e.getMessage());
+                e.printStackTrace();
```
If you want to print the stack trace, use

```java
log.error("Get broker {} path data failed which have an error", broker, e);
```

wenbingshen replied:
Thanks, I will do

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:
```diff
@@ -0,0 +1,291 @@
+        pulsarBrokers.forEach(broker -> {
+            try {
+                Optional<GetResult> brokerData = metadataStore.get(
+                        getBrokersChangePath() + "/" + broker).join();
```
`join()` throws unchecked exception, which means you don't want to catch this exception. If you want to handle the exceptional case, use `get()` and don't catch `Exception`, catch the checked exceptions.

But I also concerns the synchronous call here. Because I found `getBrokers` could also be called in `BrokersChangeEvent#process`, which is a callback. Can it be changed to asynchronous call?

BewareMyPower commented on tests/src/test/java/io/streamnative/pulsar/handlers/kop/KopEventManagerTest.java:
```diff
@@ -0,0 +1,214 @@
+        super.internalCleanup();
+    }
+
+    @Test
+    public void testGroupState() throws Exception {
```
It took about 1 minute to run this test in my local env. Is it necessary? Please and the timeout to `@Test` and reduce the test time. It's better shorter than 10 seconds even 5 seconds. If the test is unstable for shorter time, there would be some serious problems with the implementation.

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:
```diff
@@ -0,0 +1,291 @@
+    class BrokersChangeEvent implements KopEvent {
+        @Override
+        public void process() {
+            try {
+                getBrokers(metadataStore.getChildren(getBrokersChangePath()).join());
+            } catch (Exception e) {
+                log.error("BrokersChangeEvent process have an error {}", e.getMessage());
+                e.printStackTrace();
+            }
```
Synchronous call in callback is easily to cause deadlock problem. Could you change to asynchronous calls? For example,

```java
            metadataStore.getChildren(getBrokersChangePath()).whenComplete((brokers, e) -> {
                if (e != null) {
                    log.warn("BrokersChangeEvent process have an error {}", e.getMessage());
                    return;
                }
                getBrokers(brokers); // maybe it should also be async
            });
```

Or is there any reason to use the synchronous call?

wenbingshen replied:
good, you are right. I have no objection.

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:
```diff
@@ -674,11 +677,11 @@ protected void handleTopicMetadataRequest(KafkaHeaderAndRequest metadataHar,
-                        controllerId,
-                        Collections.emptyList());
+                        new MetadataResponse(
+                                Lists.newArrayList(allNodes),
```
Why did you change `allNodes` to `Lists.newArrayList(allNodes)`?

BewareMyPower replied:
Sorry it's my fault here, solve it.

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:
```diff
@@ -491,7 +492,9 @@ protected void handleTopicMetadataRequest(KafkaHeaderAndRequest metadataHar,
         // Command response for all topics
         List<TopicMetadata> allTopicMetadata = Collections.synchronizedList(Lists.newArrayList());
-        List<Node> allNodes = Collections.synchronizedList(Lists.newArrayList());
+        Set<Node> allNodes = Collections.synchronizedSet(Sets.newHashSet());
```
Why did you change `allNodes`'s type to `Set`? Is there any benefit? What I can think of is you want to remove repeated nodes. Please explain the reason.

wenbingshen replied:
Here should keep the `List`, I have changed it back. Thanks

BewareMyPower commented:
I've left some comments, PTAL. The main problem is that we should avoid waiting a future in any callback, which was very dangerous and hard to investigate the reason if deadlock happened. There was a similar problem before, see https://github.com/streamnative/kop/pull/556.

wenbingshen commented:
> I've left some comments, PTAL. The main problem is that we should avoid waiting a future in any callback, which was very dangerous and hard to investigate the reason if deadlock happened. There was a similar problem before, see #556.

Thank you for your meaningful comments. Based on your comment, I fixed the relevant code, and for the test, I added timeout=6000 ms. In my local area, the actual test can be completed in just a few hundred milliseconds. It took 1 minute before because I got the judgment condition wrong. I mistakenly wrote `describeGroup.get(groupId).state().name().equals("DEAD")` as `describeGroup.get(groupId).state( ).name().equals("Dead"))`. PTAL :)

wenbingshen commented:
A single test only took around 190 milliseconds.

![image](https://user-images.githubusercontent.com/35599757/132645123-03291d51-ac66-4bca-98a1-087da80bfb09.png)

wenbingshen commented:
The failed test does not seem to be related to the PR.

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:
```diff
@@ -0,0 +1,318 @@
+        });
+
+        kopBrokersFuture.whenComplete((brokers, e) -> {
+            if (e != null) {
+                log.error("Get pulsar brokers {} failed", pulsarBrokers, e);
+                return;
+            }
+
+            Collection<? extends Node> oldKopBrokers = adminManager.getBrokers();
+            adminManager.addBrokers(brokers);
+            log.info("Refresh kop brokers new cache {}, old brokers cache {}",
+                    adminManager.getBrokers(), oldKopBrokers);
+        });
```
`kopBrokersFuture` is not exposed, the only usage is the callback. I think we can move this part in `metadataStore.get(getBrokersChangePath() + "/" + broker)`'s callback

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/AdminManager.java:
```diff
@@ -221,4 +228,17 @@ public void shutdown() {
+        return brokersCache;
+    }
+
+    public void addBrokers(Set<Node> brokers) {
```
It's better to rename it to `setBrokers`.

BewareMyPower commented on kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadata.java:
```diff
@@ -598,6 +605,27 @@ public boolean hasPendingOffsetCommitsFromProducer(long producerId) {
         ));
     }
 
+    public Set<TopicPartition> collectPartitionsWithTopics(Set<String> topics) {
```
Maybe it's not a big issue. Since the collection only calls `addAll` method, it's better to use `List` rather than `Set`.

BewareMyPower commented:
Overall LGTM, just left a few comments, PTAL.

wenbingshen commented:
@BewareMyPower I have addressed your comments. PTAL.
